drivers/aip31068/aip31068.c

```diff
@@ -0,0 +1,516 @@
+ */
+static inline int _write(aip31068_t *dev, uint8_t data_byte, bool is_cmd);
+
+/**
+ * @brief Write data to the device.
+ *
+ * @param[in] dev   Device descriptor of the AIP31068
+ * @param[in] data  Data to write
+ * @param[in] len   Length of the data
+ *
+ * @return  AIP31068_OK on success
+ * @return  -AIP31068_ERROR_I2C if acquiring of I2C bus fails
+ * @return  -EIO When slave device doesn't ACK the byte
+ * @return  -ENXIO When no devices respond on the address sent on the bus
+ * @return  -ETIMEDOUT When timeout occurs before device's response
+ * @return  -EINVAL When an invalid argument is given
+ * @return  -EOPNOTSUPP When MCU driver doesn't support the flag operation
+ * @return  -EAGAIN When a lost bus arbitration occurs
+ */
+static int _device_write(aip31068_t *dev, uint8_t *data, uint8_t len);
+
+/**
+ * @brief Initialize controller for progress bar feature.
+ *
+ * @param[in] dev   Device descriptor of the AIP31068
+ * @param[in] row   Row where the progress bar is displayed
+ */
+static void _init_progress_bar(aip31068_t* dev, uint8_t row);
+
+/**
+ * Custom character 1 x 8 bar for progress bar.
+ */
+uint8_t custom_char_progress_bar_1[8] = { 16, 16, 16, 16, 16, 16, 16, 16 };
+
+/**
+ * Custom character 2 x 8 bar for progress bar.
+ */
+uint8_t custom_char_progress_bar_2[8] = { 24, 24, 24, 24, 24, 24, 24, 24 };
+
+/**
+ * Custom character 3 x 8 bar for progress bar.
+ */
+uint8_t custom_char_progress_bar_3[8] = { 28, 28, 28, 28, 28, 28, 28, 28 };
+
+/**
+ * Custom character 4 x 8 bar for progress bar.
+ */
+uint8_t custom_char_progress_bar_4[8] = { 30, 30, 30, 30, 30, 30, 30, 30 };
+
+/**
+ * Custom character 5 x 8 bar for progress bar.
+ */
+uint8_t custom_char_progress_bar_5[8] = { 31, 31, 31, 31, 31, 31, 31, 31 };
+
+int aip31068_init(aip31068_t *dev, const aip31068_params_t *params)
+{
+    assert(dev);
+    assert(params);
+
+    dev->params = *params;
+    dev->_curr_display_control = 0;
+    dev->_curr_entry_mode_set = 0;
+
+    i2c_init(dev->params.i2c_dev);
+
+    uint8_t _function_set = 0;
+
+    /* configure bit mode */
+    if (params->bit_mode == BITMODE_8_BIT) {
+        _function_set |= (1 << BIT_FUNCTION_SET_BITMODE);
+    }
+
+    /* configure line count */
+    if (params->row_count >= 2) {
+        _function_set |= (1 << BIT_FUNCTION_SET_LINECOUNT);
+    }
+
+    /* configure character size */
+    if (params->font_size == FONT_SIZE_5x10) {
+        _function_set |= (1 << BIT_FUNCTION_SET_FONTSIZE);
+    }
+
+    /* begin of initialization sequence (page 20 in the datasheet) */
+    xtimer_usleep(50 * US_PER_MS);
+
+    int rc = 0;
+    int c = 0;
+
+    /* send function set command sequence */
+    do {
+        rc = _command(dev, CMD_FUNCTION_SET | _function_set);
+        xtimer_usleep(5 * US_PER_MS);
+
+        printf("A rc = %d\n", rc);
+        c++;
+
+        if (c > 10) {
+            return AIP31068_ERROR_I2C;
+        }
+    } while (rc != 0); // TODO: get rid of the loop if possible
```

Comment:
Can you explain why the loop?

Reply:
I wish I knew how to get rid of this loop, but for some reason the first 8 attempts fail. If I don't loop it like this the chip seems to be stuck halfway in the initialization. The 9th attempt is successful though, no matter the initial sleep time before starting the initialization. Below the output from the loop:

```
2020-09-22 15:32:39,147 # A rc = -6
2020-09-22 15:32:39,152 # A rc = -6
2020-09-22 15:32:39,157 # A rc = -6
2020-09-22 15:32:39,163 # A rc = -6
2020-09-22 15:32:39,168 # A rc = -6
2020-09-22 15:32:39,173 # A rc = -6
2020-09-22 15:32:39,178 # A rc = -6
2020-09-22 15:32:39,183 # A rc = -6
2020-09-22 15:32:39,188 # A rc = 0
```

Maybe you have an idea?

Reply:
If you have a sequence of the syle:

```
        rc = _command(dev, CMD_FUNCTION_SET | _function_set);
        xtimer_usleep(15 * US_PER_MS); /* INCREASE THIS*/
        rc = _command(dev, CMD_FUNCTION_SET | _function_set);
```

Does it work? Or do you actually need to do the `CMD_FUNCTION_SER` multiple times in a row?

Reply:
Also is the device already initiated?

Reply:
> Or do you actually need to do the CMD_FUNCTION_SER multiple times in a row?

Well, that's actually the interesting part which I unfortunately can't answer. The data sheet of the aip31068 says that calling `CMD_FUNCTION_SET` one time is sufficient. The data sheet for the hd44780 (which is hidden behind the aip31068) says that it should be called three times. In reality I have to call it 9 times in RIOT. Interesting though is that in my Arduino version of this driver `CMD_FUNCTION_SET` is called 3 times like in the data sheet of the hd44780 and all of them return with success.
For RIOT neither changing `I2C0_SPEED` nor longer sleep periods have an effect on the first 8 calls failing.

> Also is the device already initiated?

That's the problem: It is part of the initialization^^

Reply:
What platform are you using? Does it have a different behaviour if you change `BOARD`s?

Also from your answer I'm not clear if you tried something like:

```
  rc = _command(dev, CMD_FUNCTION_SET | _function_set);
        xtimer_usleep(15 * US_PER_MS); /* INCREASE THIS*/
        rc = _command(dev, CMD_FUNCTION_SET | _function_set);
```

Reply:
> Also from your answer I'm not clear if you tried something like:

Yes I tried that with up to 500 ms sleep, but no change.

> What platform are you using?

Most of the time I'm testing with `esp32-wroom-32`. I had also tested the driver before with an `arduino-nano`, but that has been a while so I tested it again for this platform (I had to modify the test so it fits on this board). To my surprise the first call already returns successful. So it seems like there is still an issue with the esp32. In #14826 we already discussed some of my struggle but I think it's different now because it works perfectly fine for the `arduino-nano` now, but not for the esp32, even when lowering the bus speed.

I monitored the data flow with a logic analyzer. In the image below you can see the comparison between the esp32 and the Arduino nano, where the difference is basically the 8 NAKs with the esp32:

![Untitled](https://user-images.githubusercontent.com/9884864/94138732-d6155e80-fe68-11ea-9f5e-1f07dbecdf68.png)

@gschorcht Can you think of a reason for that behaviour?

Reply:
Cool, I enabled `USEMODULE += esp_i2c_hw` as you @gschorcht proposed in #14826 and this time the problem is fixed by this. So is it a valid solution to just enable this module for the esp32 when using this driver?